### PR TITLE
Patch DataLoader init on import of Lightning Lite

### LINKED ISSUE #15179: Redesign dataloader re-instantiation context managers
## Proposed refactor

Redesign the context managers that enable re-instantiations of DataLoader after #14992 deprecates the run method design.

### Motivation

We currently have two context managers over the run method:

https://github.com/Lightning-AI/lightning/blob/776432fd7e34cc02e9dbfb16ad8ec8491c176dbf/src/lightning_lite/lite.py#L394-L397

- _replace_dunder_methods(DataLoader, "dataset")
- _replace_dunder_methods(BatchSampler)

These get applied for the user if they implement the run method. However, this won't be the case anymore after #14992. We need to find an alternative approach.

### Pitch

We probably need to call these methods at the start of the program, in `LightningLite.__init__`. But it is unclear how to undo the patching which is usually done in `__exit__`. 

cc @carmocca @otaj @justusschock  

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/Lightning-AI/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/Lightning-AI/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/Lightning-AI/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/Lightning-AI/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging PyTorch Lightning, Transformers, and Hydra.

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #15179

I tested the impact the global patching has on import time.

1000 dataloaders: ~1 sec
100k dataloaders: ~3 secs
1m dataloaders: ~22 secs

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

cc @borda @justusschock @awaelchli @carmocca

### CODE REVIEW

**Review by Borda — APPROVED** [review]

File: src/lightning_lite/utilities/data.py [hunk 1]
```diff
@@ -360,6 +360,12 @@ def _replace_dunder_methods(base_cls: Type, store_explicit_arg: Optional[str] =
 
     It patches the ``__init__``, ``__setattr__`` and ``__delattr__`` methods.
     """
+    _patch_dunder_methods(base_cls, store_explicit_arg)
+    yield
+    _unpatch_dunder_methods(base_cls)
```
Comment (carmocca) [inline src/lightning_lite/utilities/data.py:365]: IMO we should remove this and go all-in into the import-level patch by having PL use it

File: tests/tests_pytorch/loops/test_loops.py [hunk 2]
```diff
@@ -28,8 +28,11 @@
 from pytorch_lightning.loops import EvaluationLoop, Loop, TrainingBatchLoop, TrainingEpochLoop
 from pytorch_lightning.trainer.progress import BaseProgress
 from pytorch_lightning.utilities.exceptions import MisconfigurationException
+from tests_pytorch.helpers.dataloaders import unpatch_dataloaders
 from tests_pytorch.helpers.runif import RunIf
 
+unpatch_dataloaders()
```
Comment (carmocca) [inline tests/tests_pytorch/loops/test_loops.py:34]: What happens if you remove this?

File: tests/tests_lite/utilities/test_data.py [hunk 3]
```diff
@@ -11,13 +14,19 @@
...
     _WrapAttrTag,
     has_iterable_dataset,
     has_len,
 )
 from lightning_lite.utilities.exceptions import MisconfigurationException
 
+# The classes already get patched on import of LightningLite
+# We need to unpatch first to have a clean starting point for all assertions in the tests below
+_unpatch_dunder_methods(DataLoader)
+_unpatch_dunder_methods(BatchSampler)
```
Comment (carmocca) [inline tests/tests_lite/utilities/test_data.py:28]: I believe this will run just once when the test files are imported (during test collection), not before the tests just in this file run.

I think the safest option would be to patch and unpatch through a conftest autouse fixture

File: tests/tests_pytorch/loops/test_loops.py (hunk 2 above)
Comment (awaelchli) [reply tests/tests_pytorch/loops/test_loops.py:34]: all tests that verify the patching fail, because

```
assert dataloader not patched  # fails here because already patched on import
with context_manager():
    assert dataloader patched
```

and similar.

Comment (carmocca) [reply tests/tests_pytorch/loops/test_loops.py:34]: I would suggest putting it at the beginning of each test that requires it then

**Comment by Borda** [comment]
@awaelchli mind checking remaining comments? :rabbit:

**Comment by Borda** [comment]
@awaelchli is this still valid for Fabric? :chipmunk:

**Comment by awaelchli** [comment]
Yes, but low priority because it only applies for users with custom dataloader subclasses. To address the concerns in the comments and issue, more work is needed. Especially since the patch needs to be applied globally, and correctly handling it in tests without interactions is quite difficult. Many tests need to be updated.
